crates/matrix-sdk-common/src/js_tracing.rs

```diff
@@ -13,86 +13,207 @@
+    logger: Option<Rc<JsLogger>>,
 
-                let _ = write!(self.string, "{value:?}");
-            }
+    /// The log level to log messages at.
+    level: Level,
+}
 
-            field_name => {
-                let _ = write!(self.string, "\n{field_name} = {value:?}");
-            }
-        }
+impl Debug for JsLogWriter {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("JsLogWriter").field("level", &self.level).finish_non_exhaustive()
     }
 }
 
-impl fmt::Display for StringVisitor {
-    fn fmt(&self, mut f: &mut fmt::Formatter<'_>) -> fmt::Result {
-        if !self.string.is_empty() {
-            write!(&mut f, " {}", self.string)
-        } else {
-            Ok(())
+impl io::Write for JsLogWriter {
+    fn write(&mut self, data: &[u8]) -> Result<usize, io::Error> {
+        use std::str;
+        let message = str::from_utf8(data)
+            .map_err(|e| io::Error::new(io::ErrorKind::InvalidData, e.to_string()))?
+            .trim_end();
+        let message = JsValue::from_str(message);
+
+        match &self.logger {
+            Some(logger) => write_message_to_logger(self.level, &message, logger.as_ref()),
+            None => write_message_to_console(self.level, &message),
         }
+
+        Ok(data.len())
+    }
+    fn flush(&mut self) -> Result<(), io::Error> {
+        Ok(())
     }
 }
+
+fn write_message_to_logger(level: Level, message: &JsValue, logger: &JsLogger) {
+    match level {
+        Level::TRACE | Level::DEBUG => logger.debug(message),
+        Level::INFO => logger.info(message),
+        Level::WARN => logger.warn(message),
+        Level::ERROR => logger.error(message),
+    };
+}
+
+fn write_message_to_console(level: Level, message: &JsValue) {
+    match level {
+        Level::TRACE | Level::DEBUG => web_sys::console::debug_1(&message),
+        Level::INFO => web_sys::console::info_1(&message),
+        Level::WARN => web_sys::console::warn_1(&message),
+        Level::ERROR => web_sys::console::error_1(&message),
+    };
+}
+
+/// The type of [`Subscriber`] returned by [`make_tracing_subscriber`]
+pub type JsLoggingSubscriber =
+    Subscriber<DefaultFields, Format<Pretty, ()>, LevelFilter, MakeJsLogWriter>;
+
+/// Construct a [`tracing::Subscriber`] which will format logs and send them to
+/// the Javascript console or the given logging object.
+///
+/// # Arguments
+///
+/// * `logger` - if `None`, logs will be sent to the JS console. Otherwise, must
+///   be a reference to a javascript object implementing `debug`, `info`, `warn`
+///   and `error` methods each taking a single `String` parameter.
+pub fn make_tracing_subscriber(logger: Option<JsLogger>) -> JsLoggingSubscriber {
+    let make_writer = match logger {
+        Some(logger) => MakeJsLogWriter::new_with_logger(logger),
+        None => MakeJsLogWriter::new(),
+    };
+
+    let format = tracing_subscriber::fmt::format().without_time().pretty();
+
+    tracing_subscriber::fmt()
```

Comment:
This fails to compile because the wasm dependency on `tracing-subscriber` is currently using `default-features = false`.